src/engine/plugins/trip.cpp

```diff
@@ -28,129 +27,151 @@ namespace engine
+    // source and end at destination by virtually squashing them together.
+    // This way the brute force and the farthest insertion algorithms don't
+    // have to be modified, and instead we can just pass a modified table to
+    // return a non-roundtrip "optimal" route from a start node to an end node.
+
+    // Original Table           // New Table
+    //   a  b  c  d  e          //   a        b         c        d         e
+    // a 0  15 36 34 30         // a 0        15        10000    34        30
+    // b 15 0  25 30 34         // b 10000    0         25       30        34
+    // c 36 25 0  18 32         // c 0        10000     0        10000     10000
+    // d 34 30 18 0  15         // d 10000    30        18       0         15
+    // e 30 34 32 15 0          // e 10000    34        32       15        0
+
+    // change parameters.source column
+    // set any node to source to impossibly high numbers so it will never
+    // try to use any node->source in the middle of the "optimal path"
+    for (std::size_t i = 0; i < result_table.GetNumberOfNodes(); i++)
+    {
+        if (i == source_id)
+            continue;
+        result_table.InvalidateRoute(i, source_id);
+    }
+
+    // change parameters.destination row
+    // set destination to anywhere else to impossibly high numbers so it will
+    // never try to use destination->any node in the middle of the "optimal path"
+    for (std::size_t i = 0; i < result_table.GetNumberOfNodes(); i++)
+    {
+        if (i == destination_id)
+            continue;
+        result_table.ShortcutRoute(destination_id, i);
+    }
+
+    // set destination->source to zero so rountrip treats source and
+    // destination as one location
+    result_table.ShortcutRoute(destination_id, source_id);
+
+    // set source->destination as very high number so algorithm is forced
+    // to find another path to get to destination
+    result_table.InvalidateRoute(source_id, destination_id);
+
+    //*********  End of changes to table  *************************************
+}
+
 Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDataFacade> facade,
                                  const api::TripParameters &parameters,
                                  util::json::Object &json_result) const
 {
     BOOST_ASSERT(parameters.IsValid());
+    const auto number_of_locations = parameters.coordinates.size();
+
+    std::size_t source_id = INVALID_SIZE_T;
+    std::size_t destination_id = INVALID_SIZE_T;
+    if (parameters.source == api::TripParameters::SourceType::First)
+    {
+        source_id = 0;
+    }
+    if (parameters.destination == api::TripParameters::DestinationType::Last)
+    {
+        BOOST_ASSERT(number_of_locations > 0);
+        destination_id = number_of_locations - 1;
+    }
+    bool fixed_start = (source_id == 0);
+    bool fixed_end = (destination_id == number_of_locations - 1);
+    bool fixed_start_and_end = fixed_start && fixed_end;
```

Comment:
It might be less confusing to not pre-compute `fixed_start_and_end`, because in the code below it was not clear if this is somehow different then `fixed_start && fixed_end`.